File: Homeworks/JS-FUNDAMENTALS/Arrays/mostFrequentNumber.js
// Write a program that finds the most frequent
// number in an array. Example:
// {4, 1, 1, 4, 2, 3, 4, 4, 1, 2, 4, 9, 3} ? 4 (5 times)

var numbers = [4, 1, 1, 4, 2, 3, 4, 4, 1, 2, 4, 9, 3];

var mode = getMode(numbers);
console.log(mode[0] + ' (' + mode[1] + ' times)');

function getMode(arr){
    var modeMap = {},
		maxNum = arr[0], 
		maxCount = 1;

    for (var ind = 0; ind < arr.length; ind++) {
        var num = arr[ind];
		
        if (modeMap[num] == null) {
            modeMap[num] = 1;
        } else {
            modeMap[num]++;
        }

        if (modeMap[num] > maxCount) {
            maxNum = num;
            maxCount = modeMap[num];
        }
    }

    return [maxNum, maxCount];
}

File: Homeworks/C#2/Arrays/MostFrequentNumber/MostFrequentNumber.cs
//Write a program that finds the most frequent number in an array. Example:
//{4, 1, 1, 4, 2, 3, 4, 4, 1, 2, 4, 9, 3}  4 (5 times)

namespace MostFrequentNumber
{
    using System;
    using System.Linq;
    class MostFrequentNumber
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Enter integers separated by a comma: ");
            int[] inputNumbers = Console.ReadLine()
                .Split(new char[] { ' ', '\t', ',' }, StringSplitOptions.RemoveEmptyEntries)
                .Select(x => int.Parse(x))
                .ToArray();

            var number = (from numbers in inputNumbers
                                      group numbers by numbers into gr
                                      orderby gr.Count() descending
                                      select new { Number = gr.Key, Frequency = gr.Count() }).First();

            Console.WriteLine("The most frequent number is {0} ({1} times)",
                number.Number, number.Frequency);
        }
    }
}
